core/src/main/java/org/testcontainers/containers/TestContainer.java

```diff
@@ -0,0 +1,194 @@
+package org.testcontainers.containers;
+
+import com.github.dockerjava.api.DockerClient;
+import com.github.dockerjava.api.command.InspectContainerResponse;
+import com.github.dockerjava.api.model.Info;
+import lombok.NonNull;
+import org.testcontainers.containers.output.OutputFrame;
+import org.testcontainers.containers.traits.*;
+import org.testcontainers.containers.wait.Wait;
+import org.testcontainers.containers.wait.WaitStrategy;
+import org.testcontainers.utility.SelfReference;
+
+import java.io.IOException;
+import java.nio.charset.Charset;
+import java.time.Duration;
+import java.util.concurrent.Future;
+import java.util.function.Consumer;
+
+public interface TestContainer<SELF extends TestContainer<SELF>> extends LinkableContainer, SelfReference<SELF>, TraitsSupport<SELF>,
+        // old commands like `withExposedPorts` are supported with Support interfaces
+        Command.Support<SELF>,
```

Comment:
I think actually my preference would be to keep the fluent-style `withX().withY().withZ()` setters for the long term. The main reason is discoverability; having all the possible operations visible through autocompletion ought to aid maintainability, especially for people who haven't read the docs or don't know to look in the traits package.

I need to have a bit more of a play with this in practice to see whether losing the explicit fluent setters hurts, but I think I'd like to keep them as the main mechanism for applying settings.

I think the Support interfaces you've set up do a good job of separating out instantiation-time from run-time concerns, and also buys some flexibility in case we do change approach in the future though.